exp/api/remote/remote_api.go

```diff
@@ -0,0 +1,530 @@
+				http.Error(w, err.Error(), http.StatusBadRequest)
+				return
+			}
+
+			// Replace the body with decompressed data
+			r.Body = io.NopCloser(bytes.NewReader(decompressed))
+			next.ServeHTTP(w, r)
+		})
+	}
+}
+
+// NewRemoteWriteHandler returns HTTP handler that receives Remote Write 2.0
+// protocol https://prometheus.io/docs/specs/remote_write_spec_2_0/.
+func NewRemoteWriteHandler(store writeStorage, opts ...HandlerOption) http.Handler {
+	o := handlerOpts{
+		logger:      slog.New(nopSlogHandler{}),
+		middlewares: []func(http.Handler) http.Handler{SnappyDecompressorMiddleware(slog.New(nopSlogHandler{}))},
+	}
+	for _, opt := range opts {
+		opt(&o)
+	}
+	h := &handler{opts: o, store: store}
+
+	// Apply all middlewares in order
+	var handler http.Handler = h
+	for i := len(o.middlewares) - 1; i >= 0; i-- {
+		handler = o.middlewares[i](handler)
+	}
+	return handler
+}
+
+// ParseProtoMsg parses the content-type header and returns the proto message type.
+//
+// The expected content-type will be of the form,
+//   - `application/x-protobuf;proto=io.prometheus.write.v2.Request` which will be treated as RW2.0 request,
+//   - `application/x-protobuf;proto=prometheus.WriteRequest` which will be treated as RW1.0 request,
+//   - `application/x-protobuf` which will be treated as RW1.0 request.
+//
+// If the content-type is not of the above forms, it will return an error.
+func ParseProtoMsg(contentType string) (WriteProtoFullName, error) {
+	contentType = strings.TrimSpace(contentType)
+
+	parts := strings.Split(contentType, ";")
+	if parts[0] != appProtoContentType {
+		return "", fmt.Errorf("expected %v as the first (media) part, got %v content-type", appProtoContentType, contentType)
+	}
+	// Parse potential https://www.rfc-editor.org/rfc/rfc9110#parameter
+	for _, p := range parts[1:] {
+		pair := strings.Split(p, "=")
+		if len(pair) != 2 {
+			return "", fmt.Errorf("as per https://www.rfc-editor.org/rfc/rfc9110#parameter expected parameters to be key-values, got %v in %v content-type", p, contentType)
+		}
+		if pair[0] == "proto" {
+			ret := WriteProtoFullName(pair[1])
+			if err := ret.Validate(); err != nil {
+				return "", fmt.Errorf("got %v content type; %w", contentType, err)
+			}
+			return ret, nil
+		}
+	}
+	// No "proto=" parameter, assuming v1.
+	return WriteProtoFullNameV1, nil
+}
+
+func (h *handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if r.Method != http.MethodPost {
+		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
+		return
+	}
+
+	contentType := r.Header.Get("Content-Type")
+	if contentType == "" {
+		contentType = appProtoContentType
+	}
+
+	msgType, err := ParseProtoMsg(contentType)
+	if err != nil {
+		h.opts.logger.Error("Error decoding remote write request", "err", err)
+		http.Error(w, err.Error(), http.StatusUnsupportedMediaType)
+		return
+	}
+
+	// Read the already decompressed body
+	body, err := io.ReadAll(r.Body)
```

Comment:
By using a `bytes.Buffer` instead of a `bytes.Reader` we could access the original bytes directly through `Bytes()` and avoid a copy/allocation here.